src/basic/macro.h

```diff
@@ -29,6 +29,19 @@
 #define _alignptr_ __attribute__((__aligned__(sizeof(void*))))
 #define _warn_unused_result_ __attribute__((__warn_unused_result__))
 
+#if defined(BUILT_WITH_COVERAGE) && BUILT_WITH_COVERAGE
+/* We need to explicitly call __gcov_dump() in places where we use _exit(), since
+ * _exit() skips at-exit hooks resulting in lost coverage */
+#  include <unistd.h>
+extern void __gcov_dump(void);
+
+_noreturn_ static inline void _coverage__exit(int status) {
+        __gcov_dump();
+        _exit(status);
+}
+#  define _exit(x) _coverage__exit(x)
+#endif
```

Comment:
Hmm, this feels dangerous. It'll only work if `macro.h` is included. If it's not included, the wrapper will not be called [1]. The problem is there will be no indication about the issue, except possibly missing coverage. So I think we need to bite the bullet and invert this: define a function with a different name that just does `_exit` if `!BUILT_WITH_COVERAGE` and replace all calls to `_exit()` with a call to it. (`do__exit()` ?)

[1]:
```console
$ git grep -n -l _exit '*.c' | xargs git grep -L macro.h
man/inotify-watch-tmp.c
man/sd-bus-container-read.c
src/analyze/analyze-exit-status.c
src/analyze/analyze.c
src/busctl/busctl.c
src/core/dbus-cgroup.c
src/core/dbus-execute.c
src/core/dbus-manager.c
src/core/dbus-path.c
src/core/dbus-scope.c
src/core/dbus-service.c
src/core/dbus-socket.c
src/core/dbus-timer.c
src/core/dbus-unit.c
src/core/dbus-util.c
src/core/load-fragment.c
src/core/main.c
src/core/mount.c
src/core/service.c
src/core/socket.c
src/core/swap.c
src/core/timer.c
src/core/unit-serialize.c
src/cryptsetup/cryptsetup.c
src/delta/delta.c
src/fsck/fsck.c
src/fuzz/fuzz-varlink.c
src/home/homectl.c
src/home/homed-home.c
src/home/homed-manager.c
src/home/homework-cifs.c
src/home/homework-fscrypt.c
src/home/homework-luks.c
src/import/export-raw.c
src/import/export-tar.c
src/import/export.c
src/import/import-common.c
src/import/import-raw.c
src/import/import-tar.c
src/import/import.c
src/import/importd.c
src/import/pull-common.c
src/import/pull.c
src/journal-remote/journal-remote-main.c
src/journal-remote/journal-upload.c
src/journal/journald-server.c
src/libsystemd-network/test-dhcp-client.c
src/libsystemd-network/test-ndisc-ra.c
src/libsystemd-network/test-ndisc-rs.c
src/libsystemd/sd-bus/bus-container.c
src/libsystemd/sd-bus/bus-control.c
src/libsystemd/sd-bus/bus-message.c
src/libsystemd/sd-bus/bus-objects.c
src/libsystemd/sd-bus/test-bus-benchmark.c
src/libsystemd/sd-bus/test-bus-marshal.c
src/libsystemd/sd-bus/test-bus-watch-bind.c
src/libsystemd/sd-netlink/netlink-genl.c
src/libsystemd/sd-netlink/netlink-message.c
```

Reply:
I'd say that our command-lines are long-enough already.

Reply:
> Hmm, this feels dangerous. It'll only work if `macro.h` is included. If it's not included, the wrapper will not be called [1]. The problem is there will be no indication about the issue, except possibly missing coverage. So I think we need to bite the bullet and invert this: define a function with a different name that just does `_exit` if `!BUILT_WITH_COVERAGE` and replace all calls to `_exit()` with a call to it. (`do__exit()` ?)

Yeah, I thought about this, but didn't want to introduce such a big change "just" because of collecting coverage. But I guess there's no better option.

Reply:
"Big" is relative — this one should be doable with a simple `sed -i` invocation, so the amount of changed lines doesn't really matter.

Reply:
I wonder - by introducing a wrapper for `_exit()` uncoditionally, we'd have to keep that in mind when introducing new code using `_exit()` (or have a Coccinelle transformation for that).

Given all this, I'm kind of inclined to the solution proposed by Jan, which would make it more "future proof" without introducing additional code changes, which are, in the end, useful for only a very specific use case (collecting coverage), thus making all this a bit more "transparent".

Reply:
Dunno, I think an approach with an explicit wrapper would be clearer. But we can try the other way too. Submit a PR and we can see how it looks.